src/cpp/rtps/reader/BaseReader.hpp

```diff
@@ -0,0 +1,398 @@
+#include <fastdds/rtps/common/EntityId_t.hpp>
+#include <fastdds/rtps/common/Guid.h>
+#include <fastdds/rtps/common/SequenceNumber.h>
+#include <fastdds/rtps/common/Time_t.h>
+#include <fastdds/rtps/common/VendorId_t.hpp>
+#include <fastdds/rtps/history/History.h>
+#include <fastdds/rtps/history/IChangePool.h>
+#include <fastdds/rtps/history/IPayloadPool.h>
+#include <fastdds/rtps/history/ReaderHistory.h>
+#include <fastdds/rtps/reader/RTPSReader.h>
+#include <fastdds/statistics/IListeners.hpp>
+#include <fastdds/statistics/rtps/StatisticsCommon.hpp>
+#include <fastdds/utils/TimedConditionVariable.hpp>
+
+namespace eprosima {
+
+namespace fastrtps {
+namespace rtps {
+
+struct CacheChange_t;
+class IDataSharingListener;
+struct ReaderHistoryState;
+class ReaderListener;
+class RTPSParticipantImpl;
+class WriterProxy;
+class WriterProxyData;
+
+} // namespace rtps
+} // namespace fastrtps
+
+namespace fastdds {
+namespace rtps {
+
+class BaseReader
+    : public fastrtps::rtps::RTPSReader
+    , public fastdds::statistics::StatisticsReaderImpl
+{
+protected:
+
+    BaseReader(
+            fastrtps::rtps::RTPSParticipantImpl* pimpl,
+            const fastrtps::rtps::GUID_t& guid,
+            const fastrtps::rtps::ReaderAttributes& att,
+            fastrtps::rtps::ReaderHistory* hist,
+            fastrtps::rtps::ReaderListener* listen);
+
+    BaseReader(
+            fastrtps::rtps::RTPSParticipantImpl* pimpl,
+            const fastrtps::rtps::GUID_t& guid,
+            const fastrtps::rtps::ReaderAttributes& att,
+            const std::shared_ptr<fastrtps::rtps::IPayloadPool>& payload_pool,
+            fastrtps::rtps::ReaderHistory* hist,
+            fastrtps::rtps::ReaderListener* listen);
+
+    BaseReader(
+            fastrtps::rtps::RTPSParticipantImpl* pimpl,
+            const fastrtps::rtps::GUID_t& guid,
+            const fastrtps::rtps::ReaderAttributes& att,
+            const std::shared_ptr<fastrtps::rtps::IPayloadPool>& payload_pool,
+            const std::shared_ptr<fastrtps::rtps::IChangePool>& change_pool,
+            fastrtps::rtps::ReaderHistory* hist,
+            fastrtps::rtps::ReaderListener* listen);
+
+public:
+
+    virtual ~BaseReader();
+
+    uint64_t get_unread_count() const override;
+
+    uint64_t get_unread_count(
+            bool mark_as_read) override;
+
+    bool wait_for_unread_cache(
+            const fastrtps::Duration_t& timeout) override;
+
+    bool is_sample_valid(
+            const void* data,
+            const fastrtps::rtps::GUID_t& writer,
+            const fastrtps::rtps::SequenceNumber_t& sn) const override;
+
+    static BaseReader* downcast(
+            fastrtps::rtps::RTPSReader* reader);
+
+    static BaseReader* downcast(
+            fastrtps::rtps::Endpoint* endpoint);
+
+    void setTrustedWriter(
+            const fastrtps::rtps::EntityId_t& writer)
+    {
+        accept_messages_from_unkown_writers_ = false;
+        m_trustedWriterEntityId = writer;
+    }
```

Comment:
Shouldn't this be protected?

Reply:
I don't think so. It is only used for builtin readers inside RTPSParticipantImpl::create_reader